### PR TITLE
chore(sdk): Add inputUri and outputUri placeholders in v1

### PR DESCRIPTION
**Pull request by chensun**
**Description of your changes:**
Add inputUri and outputUri placeholders in v1. 
- This change is based on #4719, we removed `InputUriSpec` and `OutputUriSpec` to start simple (Check commit history for the diff).

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

/assign @Ark-kun @numerology @neuromage

### CODE REVIEW

File: sdk/python/kfp/components/_components.py [hunk 1]
```diff
@@ -381,16 +381,24 @@ def component_default_to_func_default(component_default: str, is_optional: bool)
...
+    raise NotImplementedError
+
+
 def _resolve_command_line_and_paths(
     component_spec: ComponentSpec,
     arguments: Mapping[str, str],
     input_path_generator=_generate_input_file_name,
     output_path_generator=_generate_output_file_name,
     argument_serializer=serialize_value,
+    input_uri_generator=_not_implemented,
```
Comment (numerology) [inline sdk/python/kfp/components/_components.py:400]: nit: maybe type hint by the following? Similar to other parameters
```suggestion
    input_uri_generator: Callable[[str], str] = _not_implemented,
```

File: sdk/python/kfp/components/_components.py [hunk 2]
```diff
@@ -457,6 +467,26 @@ def expand_command_part(arg) -> Union[str, List[str], None]:
...
+                input_spec = inputs_dict[input_name]
+                if input_spec.optional:
+                    return None
+                else:
+                    raise ValueError('No value provided for input {}'.format(input_name))
+
+        elif isinstance(arg, OutputUriPlaceholder):
+            output_name = arg.output_name
+            output_uri = output_uri_generator(output_name)
+            output_uris[output_name] = output_uri
```
Comment (numerology) [inline sdk/python/kfp/components/_components.py:487]: Shall we check for conflicting path here, just like `OutputPathPlaceholder`?

File: sdk/python/kfp/components/_structures.py [hunk 3]
```diff
@@ -285,6 +322,12 @@ def verify_arg(arg):
                 elif isinstance(arg, OutputPathPlaceholder):
                     if arg.output_name not in self._outputs_dict:
                         raise TypeError('Argument "{}" references non-existing output.'.format(arg))
+                elif isinstance(arg, InputUriPlaceholder):
+                    if arg.input_name not in self._inputs_dict:
+                        raise TypeError('Argument "{}" references non-existing input.'.format(arg))
+                elif isinstance(arg, OutputUriPlaceholder):
```
Comment (numerology) [inline sdk/python/kfp/components/_structures.py:328]: Can we merge this with line 322? Together with the branch for `OutputPathPlaceholder`?

Similar for `InputUriPlaceholder`

**Review by numerology — COMMENTED** [review]
Thanks @chensun !

File: sdk/python/kfp/components/_components.py (hunk 1 above)
Comment (chensun) [reply sdk/python/kfp/components/_components.py:400]: Done.

File: sdk/python/kfp/components/_components.py (hunk 2 above)
Comment (chensun) [reply sdk/python/kfp/components/_components.py:487]: Done. I'll make a separate PR to do the same in v2.

File: sdk/python/kfp/components/_structures.py (hunk 3 above)
Comment (chensun) [reply sdk/python/kfp/components/_structures.py:328]: Done.

File: sdk/python/kfp/components/_components.py (hunk 2 above)
Comment (chensun) [reply sdk/python/kfp/components/_components.py]: Actually, we have the check already in v2 code.

**Review by neuromage — COMMENTED** [review]
/lgtm

**Pull request merged**